integration-cli/docker_api_containers_unix_test.go

```diff
@@ -0,0 +1,94 @@
+// +build !windows
+
+package main
+
+import (
+	"net/http"
+
+	"github.com/docker/docker/integration-cli/checker"
+	"github.com/docker/docker/integration-cli/request"
+	"github.com/docker/docker/pkg/sysinfo"
+	"github.com/go-check/check"
+)
+
+func (s *DockerSuite) TestContainersRunWithHugetlbLimit(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+	testRequires(c, SameHostDaemon)
+	testRequires(c, hugetlbLimitSupport)
+
+	sysInfo := sysinfo.New(true)
+	dSize, err := sysInfo.GetDefaultHugepageSize()
+	c.Assert(err, check.IsNil)
+
+	data := map[string]interface{}{
+		"Image":      "busybox",
+		"Cmd":        []string{"/bin/sh", "-c", "ps"},
+		"HostConfig": map[string]interface{}{"Hugetlbs": []map[string]interface{}{{"Limit": 2147483648, "PageSize": dSize}}},
+	}
+	status, resp, err := request.SockRequest("POST", "/containers/create?name=test1", data, daemonHost())
+	c.Assert(err, checker.IsNil, check.Commentf(string(resp)))
+	c.Assert(status, checker.Equals, http.StatusCreated, check.Commentf(string(resp)))
+
+	status, resp, err = request.SockRequest("POST", "/containers/test1/start", nil, daemonHost())
+	c.Assert(err, checker.IsNil)
+	c.Assert(status, checker.Equals, http.StatusNoContent)
+
+	out := inspectField(c, "test1", "HostConfig.Hugetlbs")
+	c.Assert(out, checker.Contains, "2.147483648e+09")
+}
+
+func (s *DockerSuite) TestContainersRunWithMultipleHugetlbLimit(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+	testRequires(c, SameHostDaemon)
+	testRequires(c, hugetlbLimitSupport)
+
+	sysInfo := sysinfo.New(true)
+	dSize, err := sysInfo.GetDefaultHugepageSize()
+	c.Assert(err, check.IsNil)
+
+	data := map[string]interface{}{
+		"Image":      "busybox",
+		"Cmd":        []string{"/bin/sh", "-c", "ps"},
+		"HostConfig": map[string]interface{}{"Hugetlbs": []map[string]interface{}{{"Limit": 1073741824, "PageSize": dSize}, {"Limit": 2147483648, "PageSize": dSize}}},
+	}
+	status, resp, err := request.SockRequest("POST", "/containers/create?name=test1", data, daemonHost())
+	c.Assert(err, checker.IsNil, check.Commentf(string(resp)))
+	c.Assert(status, checker.Equals, http.StatusCreated, check.Commentf(string(resp)))
+
+	status, _, err = request.SockRequest("POST", "/containers/test1/start", nil, daemonHost())
+	c.Assert(err, checker.IsNil)
+	c.Assert(status, checker.Equals, http.StatusNoContent)
+
+	// inspect has 2G setting only, 1G should be discarded
+	out := inspectField(c, "test1", "HostConfig.Hugetlbs")
```

Comment:
It would be nice to deal with the actual struct here rather than the weird inspect output.